### PR TITLE
Add traits to the `DeclarationEngine`

### PR DESCRIPTION
**Pull request by emilyaherbert**
Closes:
- #2633

Blocked by:
- #2682
- #2677

### CODE REVIEW

**Review by otrho — APPROVED** [review]
You should be able to use `.into()` instead of `CompileResult::from(...)` which is a bit less verbose.

File: sway-core/src/semantic_analysis/namespace/module.rs [hunk 1]
```diff
@@ -214,17 +214,18 @@ impl Module {
...
+            let visibility = check!(
+                decl.visibility(),
+                return err(warnings, errors),
+                warnings,
+                errors
+            );
+            if visibility == Visibility::Public {
+                symbols.push(symbol.clone());
+            }
+        }
```
[the remark this answers is not in the document]
Comment (emilyaherbert) [reply sway-core/src/semantic_analysis/namespace/module.rs:228]: > We should implement Extend or FromIterator (or something) for CompileResult

good idea!

> decide whether we want it to stop collecting on error or not though, which is often true, but not always.

It seems like the correct thing to do would probably be to create an impl that stops collecting on error (https://stackoverflow.com/a/26370894), and then figure out another solution for the other cases.

File: sway-lsp/src/capabilities/hover.rs [hunk 2]
```diff
@@ -63,7 +65,10 @@ fn hover_format(token: &Token, ident: &Ident) -> Hover {
                 TypedDeclaration::StructDeclaration(ref struct_decl) => {
                     format_visibility_hover(struct_decl.visibility, decl.friendly_name())
                 }
-                TypedDeclaration::TraitDeclaration(ref trait_decl) => {
+                TypedDeclaration::TraitDeclaration(ref decl_id) => {
+                    // TODO: do not use unwrap
```
[the remark this answers is not in the document]
Comment (emilyaherbert) [reply sway-lsp/src/capabilities/hover.rs:69]: Er, maybe ? I am not as familiar with the LSP as say, @JoshuaBatty is, so they would have more insight into the difficulty of adding error handling here. We definitely should though---internal compiler errors in this position are unlikely but still possible. My recommendation is that we use the `DocumentError::FailedToParse` error like right here: https://github.com/FuelLabs/sway/blob/3c83178581a70431971a3a824c82fd5af75afc67/sway-lsp/src/core/session.rs#L220

File: sway-core/src/semantic_analysis/namespace/module.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (otrho) [reply sway-core/src/semantic_analysis/namespace/module.rs:228]: https://github.com/FuelLabs/sway/issues/2694

**Review by otrho — APPROVED** [review]

**Pull request merged**